### PR TITLE
#724 Feed binding

### PR DESCRIPTION
**Pull request by svilenvul**
Pull request to #724

### CODE REVIEW

File: addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+        Configuration configuration = new Configuration()
+        configuration.put((FeedBindingConstants.URL), URL)
+        configuration.put((FeedBindingConstants.REFRESH_TIME),refreshTime)
+        configuration.put((FeedBindingConstants.FEED_FORMAT), format)
+        configuration.put((FeedBindingConstants.NUMBER_OF_ENTRIES), numberOfEntries)
+
+        //create Feed Thing
+        ThingUID feedUID = new ThingUID(FeedBindingConstants.FEED_THING_TYPE_UID,THING_NAME)
+        channelUID = new ChannelUID(feedUID,FeedBindingConstants.FEED_CHANNEL)
+        feedThing = managedThingProvider.createThing(FeedBindingConstants.FEED_THING_TYPE_UID, feedUID, null, null, configuration)
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy]: This method has change to "createThingofType" - please adapt your tests accordingly, thanks!

File: addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# binding
+binding.feed.name = Feed Binding
+binding.feed.description = Das Feed Binding fordert Daten von URL an, transformiert die Daten nach den Benutzeranforderungen (Format, Anzahl von Eintr�ge) und verfolgt die �nderungen in dem Feed im Zeitintervall   
+
+# thing types
+thing-type.feed.feed.label = Feed
+thing-type.feed.feed.description = Stellt zur Verf�gung Feed-Inhalt in XML Format an
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties]: "Stellt den Feed-Inhalt als XML zur Verfügung."

File: addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+thing-type.config.feed.URL.label = Feed URL
+thing-type.config.feed.URL.description = Das URL wird f�r die Lokalisierung von dem Feed benutzt
+thing-type.config.feed.refresh.label = Zeitintervall f�r Aktualisierung
+thing-type.config.feed.refresh.description = Zeitintervall f�r Aktualisierung in Minuten
+thing-type.config.feed.format.label = Ausgabe-Format
+thing-type.config.feed.format.description = Ausgabe-Format von dem Feed
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties]: "Ausgabe-Format des Feeds"

File: addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# binding
+binding.feed.name = Feed Binding
+binding.feed.description = Das Feed Binding fordert Daten von URL an, transformiert die Daten nach den Benutzeranforderungen (Format, Anzahl von Eintr�ge) und verfolgt die �nderungen in dem Feed im Zeitintervall   
+
+# thing types
+thing-type.feed.feed.label = Feed
+thing-type.feed.feed.description = Stellt zur Verf�gung Feed-Inhalt in XML Format an
+thing-type.config.feed.URL.label = Feed URL
+thing-type.config.feed.URL.description = Das URL wird f�r die Lokalisierung von dem Feed benutzt
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties]: "Die URL des Feeds"

File: addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+thing-type.config.feed.refresh.label = Zeitintervall f�r Aktualisierung
+thing-type.config.feed.refresh.description = Zeitintervall f�r Aktualisierung in Minuten
+thing-type.config.feed.format.label = Ausgabe-Format
+thing-type.config.feed.format.description = Ausgabe-Format von dem Feed
+thing-type.config.feed.length.label = Anzahl von Eintr�ge
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties]: "Anzahl der Einträge"

File: addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+thing-type.config.feed.length.description = Anzahl von gespeicherten Feeds Eintr�ge
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties]: "Die Anzahl der gespeicherten Einträge des Feeds"

File: addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+thing-type.config.feed.refresh.label = Zeitintervall f�r Aktualisierung
+thing-type.config.feed.refresh.description = Zeitintervall f�r Aktualisierung in Minuten
+thing-type.config.feed.format.label = Ausgabe-Format
+thing-type.config.feed.format.description = Ausgabe-Format von dem Feed
+thing-type.config.feed.length.label = Anzahl von Eintr�ge
+thing-type.config.feed.length.description = Anzahl von gespeicherten Feeds Eintr�ge 
+
+# channel types
+channel-type.feed.feed-channel.label = Feed Channel
+channel-type.feed.feed-channel.description = Enth�lts den aktuellen Feedinhalt
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/i18n/feed_de.properties]: "Enthält den aktuellen Inhalt des Feeds"

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+		Entry is a single element in the Feed, that contains reference (link), short description and other information like images, 
+		comments and etc. Entry in this binding is abstraction for RSS item element and Atom entry element. -->
+
+	<!-- Feed Thing Type -->
+	<thing-type id="feed">
+		<label>Feed</label>
+		<description>Provides the content of the Feed in XML format.</description>
+
+		<channels>
+			<channel id="feed-channel" typeId="feed-channel" />
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: I would give the channel a more descriptive name such as e.g. "latest-content".
Why not also provide additional channels like "no-of-entries" as a Number or "last-update" as a DateTime?

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		</channels>
+
+		<config-description>
+			<parameter name="URL" type="text" required="true">
+				<label>Feed URL</label>
+				<description>Feed URL used to localize the Feed.</description>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: why "to localize"? Isn't that THE feed url to also get all content from as well?

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+			<!--After the refresh time interval expires, the bindings checks for updates in the Feed, and if the information is not 
+				up to date, updates the feed content stored in the channel -->
+
+			<parameter name="refresh" type="integer">
+				<label>Refresh time interval</label>
+				<description>Refresh time interval in minutes.</description>
+				<default>20</default>
+			</parameter>
+
+			<!-- Feeds are converted to one of the following formats. Comparison between the two most popular formats - RSS 2.0 and
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: Hm, I think I am somehow missing the whole intention of this binding - it converts one format into another? What use is that for the user? If this is what it is doing, I would rather expect it to be a transformation service.

As a user, I would expect from such a binding to get the content (or even only the headline) of the latest post, so that I can display this in my UI or send a notification about it. Having XML within my StringItem is not really expected.

File: addons/binding/org.openhab.binding.feed/pom.xml [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
+  </parent>
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.feed</artifactId>
+  <version>2.0.0-SNAPSHOT</version>
+
+  <name>RSS Binding</name>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/pom.xml]: Everywhere else you call it "Feed Binding"

File: addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        startAutomaticRefresh();
+    }
+
+    /**
+     * This method checks if the provided configuration is valid.
+     * When invalid parameter is found, default value is assigned.
+     */
+    private void checkConfiguration() {
+        logger.debug("Start reading Feed Thing configuration.");
+        Configuration configuration = editConfiguration();
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java]: You must not edit the configuration - if you define Things in a *.things file, their configuration is read-only and cannot be changed.
Instead, use instance variables to store the values you work with (i.e. defaults if no config is set).

File: addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+            @Override
+            public void run() {
+                refreshFeedState();
+            }
+        };
+
+        refreshTask = scheduler.scheduleAtFixedRate(refresher, 0, refreshTime.intValue(), TimeUnit.MINUTES);
+    }
+
+    private void refreshFeedState() {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java]: You should also call this on a call of "channelLinked()".

File: addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+        } else {
+            logger.debug("Command {} is not supported for channel: {}. Supported command: REFRESH", command,
+                    channelUID.getId());
+        }
+    }
+
+    @Override
+    public void dispose() {
+        if (refreshTask.isDone()) {
+            logger.error("Refresh task is not terminated properly");
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java]: Isn't it perfectly valid that the job might just be running and is not yet done? If so, you should not log an error.

File: addons/binding/pom.xml [hunk 15]
```diff
@@ -20,6 +20,8 @@
     <module>org.openhab.binding.autelis</module>
     <module>org.openhab.binding.avmfritz</module>
     <module>org.openhab.binding.dscalarm</module>
+	<module>org.openhab.binding.feed</module>
```
Comment (kaikreuzer) [inline addons/binding/pom.xml]: could you please change tabs to spaces?

**Comment by kaikreuzer** [comment]
Hi @svilenvul,

Many thanks for your contribution and sorry that I took so long to look at it!
Code style looks pretty good, my main concern is about the provided functionality - I actually think that it would be more useful to extract the content and provide it through the channels (title, content, author, publishingDate, number of entries in the feed, etc.). What do you think about that?

**Comment by svilenvul** [comment]
Hi @kaikreuzer, thanks for the comments !

> Hm, I think I am somehow missing the whole intention of this binding - it converts one format into another? What use is that for the user? If this is what it is doing, I would rather expect it to be a transformation service.

> As a user, I would expect from such a binding to get the content (or even only the headline) of the latest post, so that I can display this in my UI or send a notification about it. Having XML within my StringItem is not really expected.

I am working now on refactoring the feed binding. I would like to divide it into two bundles - a transformation service and a binding.
The binding can be used for downloading, it will display last entry on the UI and and push notifications, if new entries are available. It can work independently, without the transformation.

I think the transformation service might work good in combination with XPath service - you can take any feed (any RSS or Atom format), transform it to desired format and easily extract data with XPath. 

Let me know, if you think it’s worth it :)

**Comment by kaikreuzer** [comment]
Hi @svilenvul, yes, this sounds like a good approach to me! Thanks, and sorry for making you a lot of additional work by my remarks ;-)

**Comment by svilenvul** [comment]
> Hi @svilenvul, yes, this sounds like a good approach to me! Thanks, and sorry for making you a lot of additional work by my remarks ;-)

@kaikreuzer your remarks help me to learn a lot of new things for Eclipse SmartHome and OpenHAB, so they are always welcome :). I appreciate the time that you spend on these projects.

File: addons/binding/org.openhab.binding.feed.test/pom.xml [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+	<name>Feed Binding Tests</name>
+	<packaging>eclipse-test-plugin</packaging>
+
+	<properties>
+		<maven.compiler.version>3.1</maven.compiler.version>
+		<groovy.eclipse.compiler.version>2.8.0-01</groovy.eclipse.compiler.version>
+		<groovy.eclipse.batch.version>2.1.5-03</groovy.eclipse.batch.version>
+		<build.helper.maven.plugin.version>1.8</build.helper.maven.plugin.version>
+	</properties>
+	<repositories>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/pom.xml]: are these repos really necessary? I would have hoped that everything is available in the normal target platform already?

File: addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+    @Test
+    public void 'assert that invalid configuration falls back to default values'() {
+        def String mockServletURL = generateURLString(MOCK_SERVLET_PROTOCOL,MOCK_SERVLET_HOSTNAME,MOCK_SERVLET_PORT,MOCK_SERVLET_PATH)
+        def defaultTestRefreshInterval = new BigDecimal(-10)
+        initializeFeedHandler(mockServletURL,defaultTestRefreshInterval)
+    }
+
+    @Test
+    public void 'assert that item\'s state is updated on refresh command if content changed' () {
+        boolean commandRecevied = true;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy]: typo: Received

File: addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+        boolean contentChanged = true;
+        testIfItemStateIsUpdated(commandRecevied,contentChanged);
+    }
+
+    @Test
+    public void 'assert that item\'s state is not updated on refresh command if content is not changed' () {
+        boolean commandRecevied = true;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy]: dito

File: addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy [hunk 19]
```diff
@@ -0,0 +1,401 @@
...
+        boolean contentChanged = false;
+        testIfItemStateIsUpdated(commandRecevied,contentChanged);
+    }
+
+    @Category(SlowTests.class)
+    @Test
+    public void 'assert that item\'s state is not updated on auto refresh if content is not changed' () {
+        boolean commandRecevied = false;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy]: dito

File: addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+        boolean contentChanged = false;
+        testIfItemStateIsUpdated(commandRecevied,contentChanged);
+    }
+
+    @Category(SlowTests.class)
+    @Test
+    public void 'assert that item\'s state is updated on auto refresh if content changed' () {
+        boolean commandRecevied = false;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/src/test/groovy/org/openhab/binding/feed/test/FeedHandlerTest.groovy]: dito

File: addons/binding/org.openhab.binding.feed.test/src/test/java/org/openhab/binding/feed/test/EmptyTest.java [hunk 21]
```diff
@@ -0,0 +1,12 @@
+/**
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/src/test/java/org/openhab/binding/feed/test/EmptyTest.java]: Why is this required?
Aren't the tests working if you don't have at least one Java test along the Groovy tests?

In ESH, we have [test fragments without any Java tests](https://github.com/eclipse/smarthome/tree/master/bundles/core/org.eclipse.smarthome.core.binding.xml.test/src/test), so I would assume you do not need this.

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+				<default>20</default>
+			</parameter>
+
+		</config-description>
+	</thing-type>
+
+	<channel-type id="latest-entry">
+		<item-type>String</item-type>
+		<label>Latest entry</label>
+		<description>Contains title, description and published date for the last feed entry.</description>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: This channel is not really usable as it is a concatenated string.
The "Latest Title" is something that is probably the most useful feature of the binding as it allows to include some news headline in your sitemap. I'd therefore suggest to split this channel into latest-title, latest-description, latest-published.

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+	</channel-type>
+
+	<channel-type id="author">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: I'd define this as advanced=true

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+		<item-type>String</item-type>
+		<label>Author</label>
+		<description>The name of the feed author, if author is present</description>
+	</channel-type>
+
+	<channel-type id="title">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: I'd define this as advanced=true

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+		<item-type>String</item-type>
+		<label>Title</label>
+		<description>The title of the feed</description>
+	</channel-type>
+
+	<channel-type id="description">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: I'd define this as advanced=true

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+		<item-type>String</item-type>
+		<label>Description</label>
+		<description>Description of the feed</description>
+	</channel-type>
+
+	<channel-type id="last-update">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: In what respect is this different to latest-plublished? Is this channel worthwhile?

Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: Ok, I see that feeds provide this information, so ok for me to keep it, but it should imho also be marked advanced.

File: addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+		<item-type>DateTime</item-type>
+		<label>Last update</label>
+		<description>The last update date of the feed</description>
+	</channel-type>
+
+	<channel-type id="number-of-entries">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/ESH-INF/thing/thing-types.xml]: I'd define this as advanced=true

File: addons/binding/org.openhab.binding.feed/META-INF/MANIFEST.MF [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+ org.eclipse.smarthome.core.thing.binding,
+ org.eclipse.smarthome.core.types,
+ org.openhab.binding.feed,
+ org.openhab.binding.feed.handler,
+ org.osgi.framework,
+ org.slf4j
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.openhab.binding.feed,
+ org.openhab.binding.feed.handler,
+ com.rometools.rome.io,
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/META-INF/MANIFEST.MF]: Why do you export the rome library? I assume this is required for the transformation service?
Imho, the transformation service should not depend on the presence of the binding, so you might rather want to move the lib to its own bundle (org.openhab.io.transport.feed?), which can then be used by the two add-ons. Ok for me to do this as a second step in the context of #937.

File: addons/binding/org.openhab.binding.feed/README.md [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+| title | String       | The title of the feed |
+| description | String       | Description of the feed |
+| last-update | DateTime       | The last update date of the feed |
+| number-of-entries | Number       | Number of entries in the feed |
+
+## Full Example
+
+Things:
+```
+feed:feed:news  [ URL="http://24news.com/news/rss.xml?edition=int"]
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/README.md]: Hm, both those urls do not exist. Did you make them up or did they disappear meanwhile?
It would be good to have working urls in the example!

File: addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java [hunk 30]
```diff
@@ -0,0 +1,290 @@
...
+            logger.warn("Refresh time [{}] is not valid. Falling back to default value: {}. {}", refreshTime,
+                    DEFAULT_REFRESH_TIME, e.getMessage());
+            refreshTime = DEFAULT_REFRESH_TIME;
+        }
+
+    }
+
+    private void startAutomaticRefresh() {
+
+        Runnable refresher = new Runnable() {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java]: This method is called on every channelLinked() call, of which there can be many.
You must check that there isn't already a refresh job scheduled on subsequent calls.

File: addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java [hunk 31]
```diff
@@ -0,0 +1,290 @@
...
+
+    public FeedHandler(Thing thing) {
+        super(thing);
+        currentFeedState = null;
+    }
+
+    @Override
+    public void initialize() {
+        checkConfiguration();
+        super.initialize();
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java]: I'd think you should call startAutomaticRefresh() already here - to make sure that the Thing is working and REALLY online, because maybe you cannot at all access it.

File: addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java [hunk 32]
```diff
@@ -0,0 +1,290 @@
...
+            updateFeedIfChanged(feed);
+            publishChannelIfLinked(channelUID);
+        } else {
+            logger.debug("Command {} is not supported for channel: {}. Supported command: REFRESH", command,
+                    channelUID.getId());
+        }
+    }
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java]: I think you can remove this method - the default implementation should be fine, which triggers a REFRESH command for this channel.

File: addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java [hunk 33]
```diff
@@ -0,0 +1,290 @@
...
+        } else {
+            logger.debug("No entries found");
+            return new String();
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            SyndFeed feed = fetchFeedData(urlString);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed/src/main/java/org/openhab/binding/feed/handler/FeedHandler.java]: you might want to add some safeguard here as you might receive many REFRESH commands for the different channels in a row - it can be a problem if you refresh the feed on every call. So maybe check whether you have done an update in the past 5 seconds and only refresh, if it is already older than that?

File: addons/binding/pom.xml (hunk 15 above)
Comment (kaikreuzer) [inline addons/binding/pom.xml]: please replace the tabs by spaces

Comment (kaikreuzer) [inline addons/binding/pom.xml]: Could you please add the binding also as a Karaf feature in https://github.com/openhab/openhab2-addons/blob/master/features/openhab-addons/src/main/feature/feature.xml?

**Comment by kaikreuzer** [comment]
Thanks @svilenvul, again sorry for taking so long for the review.
Please let me know if you have any questions regarding my comments.

File: addons/binding/org.openhab.binding.feed.test/pom.xml (hunk 16 above)
Comment (svilenvul) [inline addons/binding/org.openhab.binding.feed.test/pom.xml]: When I remove them , I receive error:

> [ERROR] Cannot resolve project dependencies:
[ERROR]   You requested to install 'bundle org.eclipse.equinox.http.jetty 0.0.0' but it could not be found
[ERROR]

Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.feed.test/pom.xml]: Ok, this explains the Jetty repo, but not the Groovy repo.

**Comment by svilenvul** [comment]
@kaikreuzer 
I am ready with the latest changes.

File: addons/binding/org.openhab.binding.feed.test/pom.xml (hunk 16 above)
Comment (maggu2810) [inline addons/binding/org.openhab.binding.feed.test/pom.xml]: If I open the targetplatform file in the IDE, it shows that the bundle org.eclipse.equinox.http.jetty is already part of the target platform. So, I don't know why we have to add another Jetty repo.

**Comment by kaikreuzer** [comment]
Many thanks, @svilenvul!
